overlord/devicestate/handlers_systems.go

```diff
@@ -121,6 +132,154 @@ func purgeNewSystemSnapFiles(logfile string) error {
+		}
+
+		if err := sd.LoadAssertions(nil, func(*asserts.Batch) error {
+			return nil
+		}); err != nil {
+			return nil, err
+		}
+
+		if err := sd.LoadMeta(seed.AllModes, nil, timings.New(nil)); err != nil {
+			return nil, err
+		}
+
+		err = sd.Iter(func(sn *seed.Snap) error {
+			if sn.ID() != "" {
+				requiredByOtherSystems[sn.Path] = true
+			}
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	targetSeed, err := seed.Open(dirs.SnapSeedDir, target)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := targetSeed.LoadAssertions(nil, func(*asserts.Batch) error {
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	if err := targetSeed.LoadMeta(seed.AllModes, nil, timings.New(nil)); err != nil {
+		return nil, err
+	}
+
+	var uniqueToTarget []string
+	err = targetSeed.Iter(func(sn *seed.Snap) error {
+		if sn.ID() != "" && !requiredByOtherSystems[sn.Path] {
+			uniqueToTarget = append(uniqueToTarget, sn.Path)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return uniqueToTarget, nil
+}
+
+func (m *DeviceManager) doRemoveRecoverySystem(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	systems, err := m.systems()
+	if err != nil {
+		return fmt.Errorf("cannot get recovery systems: %w", err)
+	}
+
+	setup, err := taskRemoveRecoverySystemSetup(t)
+	if err != nil {
+		return err
+	}
+
+	found := false
+	for _, sys := range systems {
+		if sys.Label == setup.Label {
+			found = true
+			if sys.Current {
+				return fmt.Errorf("cannot remove current recovery system: %q", setup.Label)
+			}
+			break
+		}
+	}
+
+	// if we couldn't find the system in the returned list of systems, then we
+	// might have already attempted to remove it. in that case, we should do our
+	// best effort at making sure that it is fully removed.
+
+	if found && len(systems) == 1 {
+		return fmt.Errorf("cannot remove last recovery system: %q", setup.Label)
+	}
+
+	deviceCtx, err := DeviceCtx(st, t, nil)
+	if err != nil {
+		return fmt.Errorf("cannot get device context: %w", err)
+	}
+
+	// if this task has partially run before, then we might have already removed
+	// some files required to load the being-removed system seed. to avoid this,
+	// we first check if we already have stored a list of snaps to remove
+	// (meaning, this task is being re-run). if the list isn't present, then we
+	// calculate it and store it in the task state for potential future re-runs.
+	var snapsToRemove uniqueSnapsInRecoverySystem
+	if err := t.Get("snaps-to-remove", &snapsToRemove); err != nil {
+		if !errors.Is(err, state.ErrNoState) {
+			return fmt.Errorf("cannot get snaps to remove from task: %w", err)
+		}
+
+		uniqueSnapPaths, err := snapsUniqueToRecoverySystem(setup.Label, systems)
+		if err != nil {
+			return fmt.Errorf("cannot get snaps unique to recovery system %q: %w", setup.Label, err)
+		}
+
+		snapsToRemove.SnapPaths = uniqueSnapPaths
+
+		t.Set("snaps-to-remove", snapsToRemove)
```

Comment:
this Set will only become effective on unlock, so a bit unsure whether it helps with the re-run problem described we would need to unlock and relock to make the data available

Reply:
Aah, I forgot about this. Given that the change would conflict with any other change that could alter the state of the seed, this should be fine I think. Though this may impede using more fine grained conflict checking if we were to address the TODO in conflict.go.